superset-frontend/src/dashboard/components/menu/DownloadMenuItems/DownloadScreenshot.tsx

```diff
@@ -0,0 +1,141 @@
+  text,
+  logEvent,
+  dashboardId,
+  addDangerToast,
+  format,
+  addSuccessToast,
+  addInfoToast,
+  ...rest
+}: {
+  text: string;
+  addDangerToast: Function;
+  dashboardId: string;
+  logEvent?: Function;
+  format: string;
+  addSuccessToast: Function;
+  addInfoToast: Function;
+}) {
+  const dashboardState = useSelector(
+    (state: RootState) => state.dashboardState,
+  );
+  const anchor = dashboardState.directPathToChild?.pop();
+  const onDownloadScreenshot = () => {
+    let retries = 0;
+
+    // this function checks if the image is ready
+    const checkImageReady = (imageUrl: string) =>
+      fetch(`${imageUrl}?download_format=${format}`)
+        .then(response => response.blob())
+        .then(blob => {
+          const url = window.URL.createObjectURL(blob);
+          const a = document.createElement('a');
+          a.href = url;
+          a.download = `screenshot.${format}`;
+          document.body.appendChild(a);
+          a.click();
+          document.body.removeChild(a);
+          window.URL.revokeObjectURL(url);
+        })
+        .catch(() => {
+          throw new Error('Image not ready');
+        });
+
+    // this is the functions that handles the retries
+    const fetchImageWithRetry = (imageUrl: string) => {
+      checkImageReady(imageUrl)
+        .then(() => {
+          addSuccessToast(t('The screenshot is now ready to be downloaded.'));
+        })
+        .catch(error => {
+          // we check how many retries have been made
+          if (retries < MAX_RETRIES) {
+            retries += 1;
+            setTimeout(() => fetchImageWithRetry(imageUrl), RETRY_INTERVAL);
+          } else {
+            logging.error(error);
+            addDangerToast(
+              t(
+                'The screenshot could not be downloaded. Please, try again later.',
+              ),
+            );
+          }
+        });
+    };
+
+    SupersetClient.post({
+      endpoint: `/api/v1/dashboard/${dashboardId}/cache_screenshot`,
+      jsonPayload: {
+        anchor,
+      },
+    })
+      .then(({ json }) => {
+        const imageUrl = json?.image_url;
+        if (imageUrl) {
+          addInfoToast(
+            'The screenshot is being generated. Please, do not leave the page.',
+          );
+        } else {
+          throw new Error('No image URL in response');
+        }
+        fetchImageWithRetry(imageUrl);
+      })
+      .catch(error => {
+        logging.error(error);
+        addDangerToast(
+          t('The screenshot could not be downloaded. Please, try again later.'),
+        );
+      })
+      .finally(() => {
+        logEvent?.(
+          format === DownloadScreenshotFormat.PNG
+            ? LOG_ACTIONS_DASHBOARD_DOWNLOAD_AS_IMAGE
+            : LOG_ACTIONS_DASHBOARD_DOWNLOAD_AS_PDF,
+        );
+      });
+  };
+
+  const item_key =
+    format === DownloadScreenshotFormat.PDF ? 'download-pdf' : 'download-image';
+
+  return (
+    <Menu.Item key={item_key} {...rest}>
```

Comment:
Can't we do `key={format}` instead of that conditional above?

Reply:
I think `DownloadScreenshotFormat.PDF` and `DownloadScreenshotFormat.PNG` should be treated as strings? And if not, `DownloadScreenshotFormat[format]` could probably work too